airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java

```diff
@@ -0,0 +1,86 @@
+import io.airbyte.protocol.models.v0.StreamDescriptor;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MssqlCursorBasedStateManager extends StreamStateManager {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MssqlCursorBasedStateManager.class);
+
+  public MssqlCursorBasedStateManager(final List<AirbyteStateMessage> airbyteStateMessages, final ConfiguredAirbyteCatalog catalog) {
+    super(airbyteStateMessages, catalog);
+  }
+
+  @Override
+  public AirbyteStateMessage toState(final Optional<AirbyteStreamNameNamespacePair> pair) {
+    if (pair.isPresent()) {
+      final Map<AirbyteStreamNameNamespacePair, CursorInfo> pairToCursorInfoMap = getPairToCursorInfoMap();
+      final Optional<CursorInfo> cursorInfo = Optional.ofNullable(pairToCursorInfoMap.get(pair.get()));
+
+      if (cursorInfo.isPresent()) {
+        LOGGER.debug("Generating state message for {}...", pair);
+        return new AirbyteStateMessage()
+            .withType(AirbyteStateType.STREAM)
+            // Temporarily include legacy state for backwards compatibility with the platform
+            .withStream(generateStreamState(pair.get(), cursorInfo.get()));
+      } else {
+        LOGGER.warn("Cursor information could not be located in state for stream {}.  Returning a new, empty state message...", pair);
+        return new AirbyteStateMessage().withType(AirbyteStateType.STREAM).withStream(new AirbyteStreamState());
+      }
+    } else {
+      LOGGER.warn("Stream not provided.  Returning a new, empty state message...");
+      return new AirbyteStateMessage().withType(AirbyteStateType.STREAM).withStream(new AirbyteStreamState());
+    }
+  }
+
+  /**
+   * Generates the stream state for the given stream and cursor information.
+   *
+   * @param airbyteStreamNameNamespacePair The stream.
+   * @param cursorInfo The current cursor.
+   * @return The {@link AirbyteStreamState} representing the current state of the stream.
+   */
+  private AirbyteStreamState generateStreamState(final AirbyteStreamNameNamespacePair airbyteStreamNameNamespacePair,
+                                                 final CursorInfo cursorInfo) {
+    return new AirbyteStreamState()
+        .withStreamDescriptor(
+            new StreamDescriptor().withName(airbyteStreamNameNamespacePair.getName()).withNamespace(airbyteStreamNameNamespacePair.getNamespace()))
+        .withStreamState(Jsons.jsonNode(generateDbStreamState(airbyteStreamNameNamespacePair, cursorInfo)));
+  }
+
+  private CursorBasedStatus generateDbStreamState(final AirbyteStreamNameNamespacePair airbyteStreamNameNamespacePair,
+                                                  final CursorInfo cursorInfo) {
+    final CursorBasedStatus state = new CursorBasedStatus();
+    state.setStateType(StateType.CURSOR_BASED);
+    state.setVersion(2L);
```

Comment:
can we hardcode this in a centralized place?

Reply:
Done